File: g5-admin/src/api/client/sftp-transfer.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { SftpTransferConcurrencyInput } from "../../types/SftpTransferConcurrencyInput";
import type { SftpTransferEnqueueInput } from "../../types/SftpTransferEnqueueInput";
import type { SftpTransferItemControlInput } from "../../types/SftpTransferItemControlInput";
import type { SftpTransferQueueSnapshot } from "../../types/SftpTransferQueueSnapshot";
import type { SftpTransferSnapshotInput } from "../../types/SftpTransferSnapshotInput";
import { invokeCommand } from "./core";

export const SFTP_TRANSFER_QUEUE_EVENT = "g5:sftp-transfer-queue";

export async function getSftpTransferQueueSnapshot(
  input: SftpTransferSnapshotInput,
): Promise<SftpTransferQueueSnapshot> {
  return invokeCommand<SftpTransferQueueSnapshot>("cmd_sftp_transfer_snapshot", { input });
}

export async function enqueueSftpTransfers(
  input: SftpTransferEnqueueInput,
): Promise<SftpTransferQueueSnapshot> {
  return invokeCommand<SftpTransferQueueSnapshot>("cmd_sftp_transfer_enqueue", { input });
}

export async function pauseSftpTransfer(
  input: SftpTransferItemControlInput,
): Promise<SftpTransferQueueSnapshot> {
  return invokeCommand<SftpTransferQueueSnapshot>("cmd_sftp_transfer_pause", { input });
}

export async function retrySftpTransfer(
  input: SftpTransferItemControlInput,
): Promise<SftpTransferQueueSnapshot> {
  return invokeCommand<SftpTransferQueueSnapshot>("cmd_sftp_transfer_retry", { input });
}

export async function cancelSftpTransfer(
  input: SftpTransferItemControlInput,
): Promise<SftpTransferQueueSnapshot> {
  return invokeCommand<SftpTransferQueueSnapshot>("cmd_sftp_transfer_cancel", { input });
}

export async function setSftpTransferConcurrency(
  input: SftpTransferConcurrencyInput,
): Promise<SftpTransferQueueSnapshot> {
  return invokeCommand<SftpTransferQueueSnapshot>("cmd_sftp_transfer_set_concurrency", { input });
}

export async function listenSftpTransferQueue(
  onSnapshot: (payload: SftpTransferQueueSnapshot) => void,
): Promise<UnlistenFn> {
  return listen<SftpTransferQueueSnapshot>(SFTP_TRANSFER_QUEUE_EVENT, ({ payload }) => {
    onSnapshot(payload);
  });
}

File: products/admin-desktop/g5-admin/src-tauri/src/commands/site/sftp_transfer.rs
use super::shared::site_command_error;
use crate::app_state::AppState;
use crate::error::CommandResult;
use g5_admin_models::models::sftp_transfer::{
    SftpTransferConcurrencyInput, SftpTransferEnqueueInput, SftpTransferItemControlInput,
    SftpTransferQueueSnapshot, SftpTransferSnapshotInput,
};
use tauri::State;

#[tauri::command]
pub async fn cmd_sftp_transfer_snapshot(
    state: State<'_, AppState>,
    input: SftpTransferSnapshotInput,
) -> CommandResult<SftpTransferQueueSnapshot> {
    state
        .sftp_transfer_service()
        .snapshot(input)
        .await
        .map_err(|error| {
            site_command_error(
                "cmd_sftp_transfer_snapshot",
                "local-sftp",
                "snapshot",
                error,
            )
        })
}

#[tauri::command]
pub async fn cmd_sftp_transfer_enqueue(
    state: State<'_, AppState>,
    input: SftpTransferEnqueueInput,
) -> CommandResult<SftpTransferQueueSnapshot> {
    state
        .sftp_transfer_service()
        .enqueue(input)
        .await
        .map_err(|error| {
            site_command_error("cmd_sftp_transfer_enqueue", "local-sftp", "enqueue", error)
        })
}

#[tauri::command]
pub async fn cmd_sftp_transfer_pause(
    state: State<'_, AppState>,
    input: SftpTransferItemControlInput,
) -> CommandResult<SftpTransferQueueSnapshot> {
    state
        .sftp_transfer_service()
        .pause(input)
        .await
        .map_err(|error| {
            site_command_error("cmd_sftp_transfer_pause", "local-sftp", "pause", error)
        })
}

#[tauri::command]
pub async fn cmd_sftp_transfer_retry(
    state: State<'_, AppState>,
    input: SftpTransferItemControlInput,
) -> CommandResult<SftpTransferQueueSnapshot> {
    state
        .sftp_transfer_service()
        .retry(input)
        .await
        .map_err(|error| {
            site_command_error("cmd_sftp_transfer_retry", "local-sftp", "retry", error)
        })
}

#[tauri::command]
pub async fn cmd_sftp_transfer_cancel(
    state: State<'_, AppState>,
    input: SftpTransferItemControlInput,
) -> CommandResult<SftpTransferQueueSnapshot> {
    state
        .sftp_transfer_service()
        .cancel(input)
        .await
        .map_err(|error| {
            site_command_error("cmd_sftp_transfer_cancel", "local-sftp", "cancel", error)
        })
}

#[tauri::command]
pub async fn cmd_sftp_transfer_set_concurrency(
    state: State<'_, AppState>,
    input: SftpTransferConcurrencyInput,
) -> CommandResult<SftpTransferQueueSnapshot> {
    state
        .sftp_transfer_service()
        .set_concurrency_limit(input)
        .await
        .map_err(|error| {
            site_command_error(
                "cmd_sftp_transfer_set_concurrency",
                "local-sftp",
                "set-concurrency",
                error,
            )
        })
}
